src/base/compute/owl_computation_cpu_init.ml

```diff
@@ -14,321 +14,296 @@ module Make
 
   open Graph.Optimiser.Operator.Symbol
 
-  open Graph.Optimiser.Operator.Symbol.Shape.Type.Device
+  open Graph.Optimiser.Operator.Symbol.Shape.Type
+
+  module MultiMap = Owl_utils_multimap.Make(
+                        struct
+                          type t = int
+                          let compare : int -> int -> int = compare
+                        end)
 
 
   (* utility functions *)
 
-  let is_initialised x =
-    let x_val = get_value x in
-    if is_elt x then false
-    else Array.length x_val > 0
-
-
-  let make_value_from src dst =
-    let dst_shp = node_shape dst in
-    match src with
-    | Some src -> (
-        (* inherit memory from the src node *)
-        let src_val = value_to_arr (get_value src).(0) in
-        let dst_val = arr_to_value (A.reshape src_val dst_shp) in
-        set_value dst [| dst_val |];
-        set_vnode dst [| src |]
+  let is_broadcastable x =
+    match get_operator x with
+    | Pow             -> true
+    | Atan2           -> true
+    | Hypot           -> true
+    | Min2            -> true
+    | Max2            -> true
+    | Add             -> true
+    | Sub             -> true
+    | Mul             -> true
+    | Div             -> true
+    | FMA             -> true
+    | EltEqual        -> true
+    | EltNotEqual     -> true
+    | EltLess         -> true
+    | EltGreater      -> true
+    | EltLessEqual    -> true
+    | EltGreaterEqual -> true
+    | _               -> false
+
+
+  (* written to be as safe as possible, but can probably set to true more
+   * operations. *)
+  let can_overwrite_parents x =
+    match get_operator x with
```

Comment:
as far as I can tell, many operations such as pooling and conv can be set to `true`. You can try to confirm with @jzstark 

Previously, I used init_00 which allocates the new memory because the input and output shape do not match, since your new algorithm can deal with this, so these can be set to true imo, to save even more memory.

Reply:
Yes, I based most of the current classification on the previous one with `init_00`. However I think `conv` cannot overwrite its parents while it is being computed, since the result of one cell depends on the value of the adjacent cells ─ so if those were modified during the computation, the result would not be correct. `pool` could be fine but it depends in which order the computation is done. Since I don't want to rely on implementation details, I prefer leaving it `false`. The person who implemented it (@jzstark) might confirm that it can overwrite the parents though :)

Reply:
agree. @jzstark took a similar strategy while optimising underlying tensor ops, starting from general optimisation, then boils down to case by case optimisation.